### PR TITLE
TripAdvisor: parser optimizations

### PR DESCRIPTION
**Pull request by remi-dupre**
So ... During my lunch break I was investigating on why using directly [`reader.read_line(&mut buffer)`](https://docs.rs/tokio/1.8.4/tokio/io/trait.AsyncBufReadExt.html#method.read_line) instead of appending with an allocated string from [`LinesStream`](https://docs.rs/tokio-stream/0.1.8/tokio_stream/wrappers/struct.LinesStream.html) was actually slower (and not in all cases as it was faster with the photos feed but slower on regular property list). Flamegraph showed that what was taking the most amount of time in *read_line* version was UTF-8 validation, so my best bet is that in the specific case of this patch, only for the indexing of property list, some compiler optimizations were not triggered :shrug: 

With that new information I couldn't resist to attempt some optimizations:

 - keep the rewritten `split_raw_properties` with fewer copies and allocations (these operations are actually non negligible according to perf)
 - delegate UTF-8 validation to parallel threads which are already doing XML validation / parsing
 - double the number of parallel threads (main thread has now a lower load while others have more work to do, so we needed to balance things a bit)
 - enable aggressive link-time optimizations, this means that it takes much more time to compile in release mode, but I think this is a reasonable compromise as this is the job we wait for on our dev workflows

With these optimizations enabled on #71 (with photos import), the imports takes about 9 minute instead of 14 when uploading from my personal computer to dev's Elasticsearch (with Liechtenstein admins loaded), which is a 50% improvement! :tada: 

**TLDR**: I think we can say that we make Rust's community proud and that our importer can be considered to be :sparkles:blazingly fast:sparkles:, especially considering that a plain gzip decompression using `zcat PropertyList.xml.gz > /dev/null` takes 8 minutes and `zcat PropertyList.xml.gz | rg '<Property ' | wc -l` takes 10 minutes while performing a subset of what we do.

**PS**: sorry about the long text, don't feel obliged to read everything if you happen to read it backward and see this first :stuck_out_tongue_winking_eye:

### CODE REVIEW

**Review by fatal69100 — APPROVED** [review]

**Comment by fatal69100** [comment]
Your comment was welcome, otherwise I would not understand what was going on. Don't worry about long text, it gives me more insights.

**Comment by remi-dupre** [comment]
> Your comment was welcome, otherwise I would not understand what was going on. Don't worry about long text, it gives me more insights.

Also, I wanted to share this blog post from which I first learned about LTO parameters: https://deterministic.space/high-performance-rust.html
